File: packages/trading/octobot_trading/util/test_tools/exchange_data.py
#  Drakkar-Software OctoBot-Trading
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.
import dataclasses
import typing

import octobot_commons.dataclasses
import octobot_commons.enums as common_enums
import octobot_trading.exchanges


@dataclasses.dataclass
class IncompatibleAssetDetails(
    octobot_commons.dataclasses.FlexibleDataclass, octobot_commons.dataclasses.UpdatableDataclass
):
    symbol: str = ""
    updated_at: float = 0


@dataclasses.dataclass
class ExchangeAuthDetails(octobot_commons.dataclasses.FlexibleDataclass, octobot_commons.dataclasses.UpdatableDataclass):
    api_key: str = ""
    api_secret: str = ""
    api_password: str = ""
    access_token: str = ""
    exchange_type: str = ""
    sandboxed: bool = False
    broker_enabled: bool = False
    encrypted: str = ""
    exchange_account_id: typing.Union[str, None] = None # deprecated, will be deleted
    exchange_credential_id: typing.Union[str, None] = None
    incompatible_assets: typing.Union[list[IncompatibleAssetDetails], None] = dataclasses.field(default_factory=list)

    # pylint: disable=E1134
    def __post_init__(self):
        if self.incompatible_assets and isinstance(self.incompatible_assets[0], dict):
            self.incompatible_assets = (
                [IncompatibleAssetDetails.from_dict(asset) for asset in self.incompatible_assets]
                if self.incompatible_assets else []
            )


@dataclasses.dataclass
class ExchangeDetails(octobot_commons.dataclasses.FlexibleDataclass, octobot_commons.dataclasses.UpdatableDataclass):
    name: str = ""


@dataclasses.dataclass
class MarketDetails(octobot_commons.dataclasses.FlexibleDataclass, octobot_commons.dataclasses.UpdatableDataclass):
    id: str = ""
    symbol: str = ""
    details: octobot_trading.exchanges.SymbolDetails = \
        dataclasses.field(default_factory=octobot_trading.exchanges.SymbolDetails)
    time_frame: str = ""
    close: list[float] = dataclasses.field(default_factory=list)
    open: list[float] = dataclasses.field(default_factory=list)
    high: list[float] = dataclasses.field(default_factory=list)
    low: list[float] = dataclasses.field(default_factory=list)
    volume: list[float] = dataclasses.field(default_factory=list)
    time: list[float] = dataclasses.field(default_factory=list)

    def has_full_candles(self):
        return self.close and self.open and self.high and self.low and self.time

    def get_formatted_candles(self) -> list[list[float]]:
        return [
            self.format_candle(index)
            for index in range(len(self.close))
        ]

    def format_candle(self, index) -> list[float]:
        ohlcv = [0.0] * len(common_enums.PriceIndexes)
        ohlcv[common_enums.PriceIndexes.IND_PRICE_TIME.value] = self.time[index]
        ohlcv[common_enums.PriceIndexes.IND_PRICE_OPEN.value] = self.open[index]
        ohlcv[common_enums.PriceIndexes.IND_PRICE_HIGH.value] = self.high[index]
        ohlcv[common_enums.PriceIndexes.IND_PRICE_LOW.value] = self.low[index]
        ohlcv[common_enums.PriceIndexes.IND_PRICE_CLOSE.value] = self.close[index]
        ohlcv[common_enums.PriceIndexes.IND_PRICE_VOL.value] = self.volume[index]
        return ohlcv


@dataclasses.dataclass
class OrdersDetails(octobot_commons.dataclasses.FlexibleDataclass, octobot_commons.dataclasses.UpdatableDataclass):
    open_orders: list[dict] = dataclasses.field(default_factory=list)
    missing_orders: list[dict] = dataclasses.field(default_factory=list)


@dataclasses.dataclass
class PortfolioDetails(octobot_commons.dataclasses.FlexibleDataclass, octobot_commons.dataclasses.UpdatableDataclass):
    initial_value: float = 0    # value of the portfolio content (not the full_content)
    content: dict = dataclasses.field(default_factory=dict) # might be a subset of global_content
    full_content: dict = dataclasses.field(default_factory=dict)  # full exchange portfolio
    asset_values: dict[str, float] = dataclasses.field(default_factory=dict)    # unitary value of each asset in reference market


@dataclasses.dataclass
class PositionDetails(octobot_commons.dataclasses.FlexibleDataclass, octobot_commons.dataclasses.UpdatableDataclass):
    position: dict = dataclasses.field(default_factory=dict)
    contract: dict = dataclasses.field(default_factory=dict)


@dataclasses.dataclass
class ExchangeData(octobot_commons.dataclasses.MinimizableDataclass, octobot_commons.dataclasses.UpdatableDataclass):
    auth_details: ExchangeAuthDetails = dataclasses.field(default_factory=ExchangeAuthDetails)
    exchange_details: ExchangeDetails = dataclasses.field(default_factory=ExchangeDetails)
    markets: list[MarketDetails] = dataclasses.field(default_factory=list)
    portfolio_details: PortfolioDetails = dataclasses.field(default_factory=PortfolioDetails)
    orders_details: OrdersDetails = dataclasses.field(default_factory=OrdersDetails)
    trades: list[dict] = dataclasses.field(default_factory=list)
    positions: list[PositionDetails] = dataclasses.field(default_factory=list)

    # pylint: disable=E1134
    def __post_init__(self):
        if self.markets and isinstance(self.markets[0], dict):
            self.markets = [MarketDetails.from_dict(market) for market in self.markets] if self.markets else []
        if self.positions and isinstance(self.positions[0], dict):
            self.positions = [PositionDetails.from_dict(position) for position in self.positions] \
                if self.positions else []

    def get_price(self, symbol):
        for market in self.markets:
            if market.symbol == symbol:
                return market.close[-1]
        raise KeyError(symbol)

File: packages/trading/octobot-trading-ts/src/exchanges/util/exchangeData.ts
import Decimal from "decimal.js";
import { SymbolDetails, makeSymbolDetails } from "./symbolDetails.js";
import {
  PRICE_INDEX_TIME,
  PRICE_INDEX_OPEN,
  PRICE_INDEX_HIGH,
  PRICE_INDEX_LOW,
  PRICE_INDEX_CLOSE,
  PRICE_INDEX_VOL,
} from "../adapters/ccxtAdapter.js";

export interface IncompatibleAssetDetails {
  symbol: string;
  updatedAt: number;
}

export interface ExchangeAuthDetails {
  apiKey: string;
  apiSecret: string;
  apiPassword: string;
  accessToken: string;
  exchangeType: string;
  sandboxed: boolean;
  brokerEnabled: boolean;
  encrypted: string;
  exchangeAccountId: string | null;
  exchangeCredentialId: string | null;
  incompatibleAssets: IncompatibleAssetDetails[] | null;
}

export interface ExchangeDetailsData {
  name: string;
}

export class MarketDetails {
  id = "";
  symbol = "";
  details: SymbolDetails = makeSymbolDetails();
  timeFrame = "";
  close: number[] = [];
  open: number[] = [];
  high: number[] = [];
  low: number[] = [];
  volume: number[] = [];
  time: number[] = [];

  hasFullCandles(): boolean {
    return (
      this.close.length > 0 &&
      this.open.length > 0 &&
      this.high.length > 0 &&
      this.low.length > 0 &&
      this.time.length > 0
    );
  }

  formatCandle(index: number): number[] {
    const ohlcv = new Array(6).fill(0);
    ohlcv[PRICE_INDEX_TIME] = this.time[index];
    ohlcv[PRICE_INDEX_OPEN] = this.open[index];
    ohlcv[PRICE_INDEX_HIGH] = this.high[index];
    ohlcv[PRICE_INDEX_LOW] = this.low[index];
    ohlcv[PRICE_INDEX_CLOSE] = this.close[index];
    ohlcv[PRICE_INDEX_VOL] = this.volume[index];
    return ohlcv;
  }

  getFormattedCandles(): number[][] {
    return Array.from({ length: this.close.length }, (_, i) => this.formatCandle(i));
  }

  static fromOhlcvs(symbol: string, timeFrame: string, ohlcvs: number[][]): MarketDetails {
    const m = new MarketDetails();
    m.symbol = symbol;
    m.timeFrame = timeFrame;
    m.close = ohlcvs.map((o) => o[PRICE_INDEX_CLOSE]);
    m.open = ohlcvs.map((o) => o[PRICE_INDEX_OPEN]);
    m.high = ohlcvs.map((o) => o[PRICE_INDEX_HIGH]);
    m.low = ohlcvs.map((o) => o[PRICE_INDEX_LOW]);
    m.volume = ohlcvs.map((o) => o[PRICE_INDEX_VOL]);
    m.time = ohlcvs.map((o) => o[PRICE_INDEX_TIME]);
    return m;
  }
}

export interface OrdersDetails {
  openOrders: Record<string, unknown>[];
  missingOrders: Record<string, unknown>[];
}

export interface PortfolioDetails {
  initialValue: number;
  content: Record<string, Record<string, number | Decimal>>;
  fullContent: Record<string, Record<string, number | Decimal>>;
  assetValues: Record<string, number>;
}

export interface PositionDetails {
  position: Record<string, unknown>;
  contract: Record<string, unknown>;
}

export class ExchangeData {
  authDetails: ExchangeAuthDetails = {
    apiKey: "",
    apiSecret: "",
    apiPassword: "",
    accessToken: "",
    exchangeType: "",
    sandboxed: false,
    brokerEnabled: false,
    encrypted: "",
    exchangeAccountId: null,
    exchangeCredentialId: null,
    incompatibleAssets: [],
  };
  exchangeDetails: ExchangeDetailsData = { name: "" };
  markets: MarketDetails[] = [];
  portfolioDetails: PortfolioDetails = {
    initialValue: 0,
    content: {},
    fullContent: {},
    assetValues: {},
  };
  ordersDetails: OrdersDetails = { openOrders: [], missingOrders: [] };
  trades: Record<string, unknown>[] = [];
  positions: PositionDetails[] = [];

  getPrice(symbol: string): number {
    for (const market of this.markets) {
      if (market.symbol === symbol) return market.close[market.close.length - 1];
    }
    throw new Error(`Symbol not found in exchange data: ${symbol}`);
  }
}

export function exchangeDataFactory(
  exchangeInternalName: string,
  exchangeType?: string,
  sandboxed = false,
  exchangeCredentialId?: string,
  authDetails?: Partial<ExchangeAuthDetails>,
): ExchangeData {
  const data = new ExchangeData();
  data.exchangeDetails.name = exchangeInternalName;
  data.authDetails.sandboxed = sandboxed;
  data.authDetails.exchangeCredentialId = exchangeCredentialId ?? null;
  if (exchangeType) data.authDetails.exchangeType = exchangeType;
  if (authDetails) Object.assign(data.authDetails, authDetails);
  return data;
}
